File: backend/app/api/routes/targets.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
import uuid
import json
import os

from app.api.routes.auth import get_current_active_user

router = APIRouter()

# File-based storage for persistence
TARGETS_FILE = "/tmp/msf_gui_targets.json"


# Pydantic Models
class ServiceCreate(BaseModel):
    port: int
    protocol: str = "tcp"
    service: str = ""
    version: Optional[str] = None
    banner: Optional[str] = None
    state: str = "open"


class Service(ServiceCreate):
    id: str
    host_id: str
    created_at: str


class TargetCreate(BaseModel):
    ip: str
    hostname: Optional[str] = None
    os: Optional[str] = None
    os_family: Optional[str] = None
    arch: Optional[str] = None
    status: str = "unknown"  # unknown, online, offline, compromised
    tags: List[str] = []
    notes: Optional[str] = None
    group: Optional[str] = None


class TargetUpdate(BaseModel):
    ip: Optional[str] = None
    hostname: Optional[str] = None
    os: Optional[str] = None
    os_family: Optional[str] = None
    arch: Optional[str] = None
    status: Optional[str] = None
    tags: Optional[List[str]] = None
    notes: Optional[str] = None
    group: Optional[str] = None


class Target(TargetCreate):
    id: str
    created_at: str
    updated_at: str
    services: List[Service] = []
    session_count: int = 0


class TargetImport(BaseModel):
    targets: List[TargetCreate]


# In-memory storage with file persistence
targets_db: dict[str, dict] = {}
services_db: dict[str, dict] = {}


def load_targets():
    """Load targets from file."""
    global targets_db, services_db
    if os.path.exists(TARGETS_FILE):
        try:
            with open(TARGETS_FILE, 'r') as f:
                data = json.load(f)
                targets_db = data.get("targets", {})
                services_db = data.get("services", {})
        except Exception:
            targets_db = {}
            services_db = {}


def save_targets():
    """Save targets to file."""
    try:
        with open(TARGETS_FILE, 'w') as f:
            json.dump({"targets": targets_db, "services": services_db}, f, indent=2)
    except Exception as e:
        print(f"Failed to save targets: {e}")


# Load on module import
load_targets()


def get_target_with_services(target_id: str) -> dict:
    """Get target with its services attached."""
    if target_id not in targets_db:
        return None
    target = targets_db[target_id].copy()
    target["services"] = [
        s for s in services_db.values() if s.get("host_id") == target_id
    ]
    return target


# Target CRUD endpoints
@router.get("")
async def list_targets(
    status: Optional[str] = None,
    group: Optional[str] = None,
    tag: Optional[str] = None,
    user=Depends(get_current_active_user)
):
    """List all targets with optional filtering."""
    result = []
    for target_id, target in targets_db.items():
        # Apply filters
        if status and target.get("status") != status:
            continue
        if group and target.get("group") != group:
            continue
        if tag and tag not in target.get("tags", []):
            continue

        # Get full target with services
        full_target = get_target_with_services(target_id)
        if full_target:
            result.append(full_target)

    # Sort by created_at descending
    result.sort(key=lambda x: x.get("created_at", ""), reverse=True)

    return {
        "targets": result,
        "count": len(result),
        "groups": list(set(t.get("group") for t in targets_db.values() if t.get("group"))),
        "tags": list(set(tag for t in targets_db.values() for tag in t.get("tags", [])))
    }


@router.post("")
async def create_target(target: TargetCreate, user=Depends(get_current_active_user)):
    """Create a new target."""
    # Check for duplicate IP
    for existing in targets_db.values():
        if existing.get("ip") == target.ip:
            raise HTTPException(status_code=400, detail=f"Target with IP {target.ip} already exists")

    target_id = str(uuid.uuid4())
    now = datetime.now().isoformat()

    target_data = {
        "id": target_id,
        **target.model_dump(),
        "created_at": now,
        "updated_at": now,
        "session_count": 0
    }

    targets_db[target_id] = target_data
    save_targets()

    return get_target_with_services(target_id)


@router.get("/{target_id}")
async def get_target(target_id: str, user=Depends(get_current_active_user)):
    """Get a specific target by ID."""
    target = get_target_with_services(target_id)
    if not target:
        raise HTTPException(status_code=404, detail="Target not found")
    return target


@router.put("/{target_id}")
async def update_target(target_id: str, update: TargetUpdate, user=Depends(get_current_active_user)):
    """Update a target."""
    if target_id not in targets_db:
        raise HTTPException(status_code=404, detail="Target not found")

    # Check for duplicate IP if IP is being changed
    if update.ip:
        for tid, existing in targets_db.items():
            if tid != target_id and existing.get("ip") == update.ip:
                raise HTTPException(status_code=400, detail=f"Target with IP {update.ip} already exists")

    # Update only provided fields
    update_data = update.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        if value is not None:
            targets_db[target_id][key] = value

    targets_db[target_id]["updated_at"] = datetime.now().isoformat()
    save_targets()

    return get_target_with_services(target_id)


@router.delete("/{target_id}")
async def delete_target(target_id: str, user=Depends(get_current_active_user)):
    """Delete a target and its services."""
    if target_id not in targets_db:
        raise HTTPException(status_code=404, detail="Target not found")

    # Delete associated services
    services_to_delete = [sid for sid, s in services_db.items() if s.get("host_id") == target_id]
    for sid in services_to_delete:
        del services_db[sid]

    del targets_db[target_id]
    save_targets()

    return {"success": True, "message": "Target deleted"}


@router.post("/import")
async def import_targets(data: TargetImport, user=Depends(get_current_active_user)):
    """Import multiple targets at once."""
    imported = 0
    skipped = 0

    for target in data.targets:
        # Skip duplicates
        exists = any(t.get("ip") == target.ip for t in targets_db.values())
        if exists:
            skipped += 1
            continue

        target_id = str(uuid.uuid4())
        now = datetime.now().isoformat()

        target_data = {
            "id": target_id,
            **target.model_dump(),
            "created_at": now,
            "updated_at": now,
            "session_count": 0
        }

        targets_db[target_id] = target_data
        imported += 1

    save_targets()

    return {
        "success": True,
        "imported": imported,
        "skipped": skipped,
        "total": len(targets_db)
    }


# Service endpoints
@router.post("/{target_id}/services")
async def add_service(target_id: str, service: ServiceCreate, user=Depends(get_current_active_user)):
    """Add a service to a target."""
    if target_id not in targets_db:
        raise HTTPException(status_code=404, detail="Target not found")

    # Check for duplicate port/protocol
    for s in services_db.values():
        if (s.get("host_id") == target_id and
            s.get("port") == service.port and
            s.get("protocol") == service.protocol):
            raise HTTPException(status_code=400, detail=f"Service on port {service.port}/{service.protocol} already exists")

    service_id = str(uuid.uuid4())
    service_data = {
        "id": service_id,
        "host_id": target_id,
        **service.model_dump(),
        "created_at": datetime.now().isoformat()
    }

    services_db[service_id] = service_data
    save_targets()

    return service_data


@router.get("/{target_id}/services")
async def list_services(target_id: str, user=Depends(get_current_active_user)):
    """List all services for a target."""
    if target_id not in targets_db:
        raise HTTPException(status_code=404, detail="Target not found")

    services = [s for s in services_db.values() if s.get("host_id") == target_id]
    services.sort(key=lambda x: x.get("port", 0))

    return {"services": services, "count": len(services)}


@router.delete("/{target_id}/services/{service_id}")
async def delete_service(target_id: str, service_id: str, user=Depends(get_current_active_user)):
    """Delete a service from a target."""
    if target_id not in targets_db:
        raise HTTPException(status_code=404, detail="Target not found")

    if service_id not in services_db:
        raise HTTPException(status_code=404, detail="Service not found")

    if services_db[service_id].get("host_id") != target_id:
        raise HTTPException(status_code=400, detail="Service does not belong to this target")

    del services_db[service_id]
    save_targets()

    return {"success": True, "message": "Service deleted"}


# Bulk operations
@router.post("/bulk/status")
async def bulk_update_status(
    target_ids: List[str],
    status: str,
    user=Depends(get_current_active_user)
):
    """Update status for multiple targets."""
    updated = 0
    for target_id in target_ids:
        if target_id in targets_db:
            targets_db[target_id]["status"] = status
            targets_db[target_id]["updated_at"] = datetime.now().isoformat()
            updated += 1

    save_targets()
    return {"success": True, "updated": updated}


@router.delete("/bulk")
async def bulk_delete(target_ids: List[str], user=Depends(get_current_active_user)):
    """Delete multiple targets."""
    deleted = 0
    for target_id in target_ids:
        if target_id in targets_db:
            # Delete services
            services_to_delete = [sid for sid, s in services_db.items() if s.get("host_id") == target_id]
            for sid in services_to_delete:
                del services_db[sid]
            del targets_db[target_id]
            deleted += 1

    save_targets()
    return {"success": True, "deleted": deleted}


# Stats endpoint
@router.get("/stats/summary")
async def get_stats(user=Depends(get_current_active_user)):
    """Get target statistics."""
    total = len(targets_db)
    by_status = {}
    by_os = {}
    by_group = {}

    for target in targets_db.values():
        status = target.get("status", "unknown")
        by_status[status] = by_status.get(status, 0) + 1

        os_family = target.get("os_family", "unknown")
        by_os[os_family] = by_os.get(os_family, 0) + 1

        group = target.get("group", "ungrouped")
        by_group[group] = by_group.get(group, 0) + 1

    return {
        "total": total,
        "by_status": by_status,
        "by_os": by_os,
        "by_group": by_group,
        "total_services": len(services_db)
    }

File: backend_ruby/standalone/routes/targets.rb
# frozen_string_literal: true

require 'sinatra/base'
require 'securerandom'

module MsfGui
  module Routes
    class Targets < Sinatra::Base
      helpers Helpers::Response
      helpers Helpers::Validators

      configure do
        set :show_exceptions, false
      end

      # List targets with optional filters
      get '/' do
        status_filter = params[:status]
        group_filter = params[:group]
        tag_filter = params[:tag]

        data = Stores.targets.load({ targets: [] })
        targets = data[:targets] || data['targets'] || []

        # Apply filters
        if status_filter && !status_filter.empty?
          targets = targets.select { |t| (t[:status] || t['status']) == status_filter }
        end

        if group_filter && !group_filter.empty?
          targets = targets.select { |t| (t[:group] || t['group']) == group_filter }
        end

        if tag_filter && !tag_filter.empty?
          targets = targets.select do |t|
            tags = t[:tags] || t['tags'] || []
            tags.include?(tag_filter)
          end
        end

        json_response(targets)
      end

      # Create target
      post '/' do
        body = parse_json_body

        target = {
          id: SecureRandom.uuid,
          ip: body[:ip] || body['ip'],
          hostname: body[:hostname] || body['hostname'],
          os: body[:os] || body['os'],
          os_family: body[:os_family] || body['os_family'],
          arch: body[:arch] || body['arch'],
          status: body[:status] || body['status'] || 'unknown',
          tags: body[:tags] || body['tags'] || [],
          notes: body[:notes] || body['notes'],
          group: body[:group] || body['group'],
          session_count: 0,
          services: [],
          created_at: Time.now.iso8601,
          updated_at: Time.now.iso8601
        }

        unless target[:ip] || target[:hostname]
          json_error('IP or hostname is required', status: 422)
        end

        Stores.targets.update do |data|
          data[:targets] ||= []
          data[:targets] << target
          data
        end

        json_response(target, status: 201)
      end

      # Get target with services
      get '/:target_id' do
        data = Stores.targets.load({ targets: [] })
        targets = data[:targets] || data['targets'] || []
        target = targets.find { |t| (t[:id] || t['id']) == params[:target_id] }

        unless target
          json_error('Target not found', status: 404)
        end

        json_response(target)
      end

      # Update target
      put '/:target_id' do
        body = parse_json_body
        target_id = params[:target_id]

        updated = nil
        Stores.targets.update do |data|
          data[:targets] ||= []
          idx = data[:targets].find_index { |t| (t[:id] || t['id']) == target_id }

          unless idx
            json_error('Target not found', status: 404)
          end

          # Update fields
          %i[ip hostname os os_family arch status tags notes group].each do |field|
            if body[field] || body[field.to_s]
              data[:targets][idx][field] = body[field] || body[field.to_s]
            end
          end
          data[:targets][idx][:updated_at] = Time.now.iso8601
          updated = data[:targets][idx]
          data
        end

        json_response(updated)
      end

      # Delete target
      delete '/:target_id' do
        target_id = params[:target_id]

        Stores.targets.update do |data|
          data[:targets] ||= []
          data[:targets].reject! { |t| (t[:id] || t['id']) == target_id }
          data
        end

        json_response({ success: true, message: 'Target deleted' })
      end

      # Bulk import targets
      post '/import' do
        body = parse_json_body
        targets_data = body[:targets] || body['targets'] || []

        imported = []
        Stores.targets.update do |data|
          data[:targets] ||= []

          targets_data.each do |t|
            target = {
              id: SecureRandom.uuid,
              ip: t[:ip] || t['ip'],
              hostname: t[:hostname] || t['hostname'],
              os: t[:os] || t['os'],
              os_family: t[:os_family] || t['os_family'],
              arch: t[:arch] || t['arch'],
              status: t[:status] || t['status'] || 'unknown',
              tags: t[:tags] || t['tags'] || [],
              notes: t[:notes] || t['notes'],
              group: t[:group] || t['group'],
              services: t[:services] || t['services'] || [],
              session_count: 0,
              created_at: Time.now.iso8601,
              updated_at: Time.now.iso8601
            }

            next unless target[:ip] || target[:hostname]

            data[:targets] << target
            imported << target
          end

          data
        end

        json_response({
          success: true,
          imported: imported.size,
          targets: imported
        })
      end

      # Get services for target
      get '/:target_id/services' do
        data = Stores.targets.load({ targets: [] })
        targets = data[:targets] || data['targets'] || []
        target = targets.find { |t| (t[:id] || t['id']) == params[:target_id] }

        unless target
          json_error('Target not found', status: 404)
        end

        services = target[:services] || target['services'] || []
        json_response(services)
      end

      # Add service to target
      post '/:target_id/services' do
        body = parse_json_body
        target_id = params[:target_id]

        service = {
          id: SecureRandom.uuid,
          port: body[:port] || body['port'],
          protocol: body[:protocol] || body['protocol'] || 'tcp',
          name: body[:name] || body['name'],
          version: body[:version] || body['version'],
          state: body[:state] || body['state'] || 'open',
          banner: body[:banner] || body['banner'],
          created_at: Time.now.iso8601
        }

        unless service[:port]
          json_error('Port is required', status: 422)
        end

        Stores.targets.update do |data|
          data[:targets] ||= []
          idx = data[:targets].find_index { |t| (t[:id] || t['id']) == target_id }

          unless idx
            json_error('Target not found', status: 404)
          end

          data[:targets][idx][:services] ||= []
          data[:targets][idx][:services] << service
          data[:targets][idx][:updated_at] = Time.now.iso8601
          data
        end

        json_response(service, status: 201)
      end

      # Delete service from target
      delete '/:target_id/services/:service_id' do
        target_id = params[:target_id]
        service_id = params[:service_id]

        Stores.targets.update do |data|
          data[:targets] ||= []
          idx = data[:targets].find_index { |t| (t[:id] || t['id']) == target_id }

          unless idx
            json_error('Target not found', status: 404)
          end

          data[:targets][idx][:services] ||= []
          data[:targets][idx][:services].reject! { |s| (s[:id] || s['id']) == service_id }
          data[:targets][idx][:updated_at] = Time.now.iso8601
          data
        end

        json_response({ success: true, message: 'Service deleted' })
      end

      # Bulk update status
      post '/bulk/status' do
        body = parse_json_body
        target_ids = body[:ids] || body['ids'] || []
        new_status = body[:status] || body['status']

        unless new_status
          json_error('Status is required', status: 422)
        end

        updated = 0
        Stores.targets.update do |data|
          data[:targets] ||= []
          data[:targets].each do |target|
            if target_ids.include?(target[:id] || target['id'])
              target[:status] = new_status
              target[:updated_at] = Time.now.iso8601
              updated += 1
            end
          end
          data
        end

        json_response({ success: true, updated: updated })
      end

      # Bulk delete
      delete '/bulk' do
        body = parse_json_body
        target_ids = body[:ids] || body['ids'] || []

        deleted = 0
        Stores.targets.update do |data|
          data[:targets] ||= []
          original_count = data[:targets].size
          data[:targets].reject! { |t| target_ids.include?(t[:id] || t['id']) }
          deleted = original_count - data[:targets].size
          data
        end

        json_response({ success: true, deleted: deleted })
      end

      # Get target statistics
      get '/stats/summary' do
        data = Stores.targets.load({ targets: [] })
        targets = data[:targets] || data['targets'] || []

        # Count by status
        status_counts = targets.group_by { |t| t[:status] || t['status'] }
                               .transform_values(&:size)

        # Count by OS family
        os_counts = targets.group_by { |t| t[:os_family] || t['os_family'] || 'unknown' }
                           .transform_values(&:size)

        # Count by group
        group_counts = targets.group_by { |t| t[:group] || t['group'] || 'ungrouped' }
                              .transform_values(&:size)

        json_response({
          total: targets.size,
          by_status: status_counts,
          by_os: os_counts,
          by_group: group_counts
        })
      end
    end
  end
end
